### PR TITLE
Polynomial multiplication, Default trait

### PR DESCRIPTION
**Pull request by shargs**
- Polynomial multiplication mod X^n + 1
- Reorganisation of algebraic traits
- Fixed Shake
- Algorithms 1, 2, 3
- Functions H, G, XOF, KDF
- Basic RNG
- CCAKEM keygen and encryption
- Compress / Decompress

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -91,64 +108,135 @@ pub fn kyber_ccakem_dec(_c: &ByteArray, _sk: &ByteArray) -> ByteArray {
...
+    let mut i = 0;
+    let mut j = 0;
+    let mut coeffs = vec![F3329::default(); n];
+    while j < n {
+        let d = (bs.data[i] as usize) + (bs.data[i + 1] as usize) << 8;
+        if d < 19 * q {
+            coeffs[j] = F3329::from_int(d.try_into().unwrap());
+            j += 1;
+        }
+        i += 2;
```
Comment (ederoffi) [inline src/lib.rs]: keeps increasing but no check for index overflow

File: src/lib.rs [hunk 2]
```diff
@@ -91,64 +108,135 @@ pub fn kyber_ccakem_dec(_c: &ByteArray, _sk: &ByteArray) -> ByteArray {
...
+    let input = r.clone().append(&j_as_bytes).append(&i_as_bytes);
+    ByteArray {
+        data: hash::shake_128(input.data, len),
+    }
 }
 
 // Hash function => SHA3-256
-fn h(_r: ByteArray) -> (ByteArray, ByteArray) {
-    unimplemented!();
+fn h(r: &ByteArray) -> (ByteArray, ByteArray) {
```
Comment (ederoffi) [inline src/lib.rs]: `h` should be consistent with `g` and both should either take a ref or a value

Comment (ederoffi) [reply src/lib.rs]: Should it be true for other function like `xof` ?

File: src/polyvec/polyvec.rs [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
-use std::ops::{Index, IndexMut};
+use crate::polyvec::structures::{FiniteRing, RingModule};
 
 #[derive(Clone)]
-pub struct Polyvec<T: FiniteRing> {
-    coefficients: Vec<T>,
-    dimension: usize,
+pub struct PolyVec<T: FiniteRing> {
+    pub coefficients: Vec<T>,
+    pub dimension: usize,
```
Comment (ederoffi) [inline src/polyvec/polyvec.rs:14]: why pub since we have getters and setters ?

File: src/lib.rs (hunk 1 above)
Comment (shargs) [reply src/lib.rs]: Probability of overflow can be computed and is negligeably slow

File: src/lib.rs (hunk 2 above)
Comment (shargs) [reply src/lib.rs]: Both should take a ref; changes were implemented as they seemed necessary, eventually when reaching g the same modification would have been made

File: src/polyvec/polyvec.rs (hunk 3 above)
Comment (shargs) [reply src/polyvec/polyvec.rs:14]: There is no secret data so setters/getters in this case seem overkill, plus exposing `coefficients` provides iterators and item which makes it more practical to use

File: src/bytearray.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    pub fn concat(items: &[&Self]) -> Self {
+        if items.len() == 0 {
+            return Self { data: vec![] };
+        }
+
+        let mut data = items.first().unwrap().data.clone();
+        for item in items.iter().skip(1) {
+            data.extend(item.data.iter().cloned());
```
Comment (a1phyr) [inline src/bytearray.rs]: I think you should prefer using [`Vec::extend_from_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice) here.

File: src/bytearray.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        for (i, c) in self.data.iter().enumerate() {
+            if other.data[i] != *c {
+                return false;
+            }
+        }
+        true
+    }
+}
+impl Eq for ByteArray {}
```
Comment (a1phyr) [inline src/bytearray.rs]: `#[derive(PartialEq, Eq)]` on `ByteArray` would be more idiomatic and more efficient.

File: src/polyvec/polyvec.rs [hunk 6]
```diff
@@ -25,12 +36,31 @@ where
...
+    fn is_zero(&self) -> bool {
+        if self.dimension == 0 {
+            true
+        } else {
+            for c in self.coefficients.iter() {
+                if !c.is_zero() {
+                    return false;
+                }
+            }
+            true
```
Comment (a1phyr) [inline src/polyvec/polyvec.rs]: I think you should better use `self.coefficients.iter().all(|c| c.is_zero())`

File: src/lib.rs (hunk 2 above)
Comment (a1phyr) [reply src/lib.rs]: Is there a specific reason they should take a ref ? Given, the current implementation, taking the array by value would avoid the clone.

**Comment by ederoffi** [comment]
Test are now all green
I suggest we now merge this huge PR and move on to cleanup a new cleanup or optimisation one

**Review by ederoffi — APPROVED** [review]

**Pull request merged**